docs/hooks/new-language.rst

```diff
@@ -1,105 +1,120 @@
 
-When you run Dredd with :option:`--language` option, it runs the given command and tries to connect to ``http://127.0.0.1:61321``. If connection to the hook handling server wasn’t successful, it exits with exit code ``3``.
+.. note::
+   `Deserve eternal praise <hall-of-fame>`__ and contribute hooks handler for **Java**! See :ghissue:`#875`
 
-Dredd internally registers a function for each :ref:`type of hooks <types-of-hooks>` and when this function is executed it assigns execution ``uuid`` to that event, serializes received function parameters (a :ref:`Transaction object <transaction>` or an Array of it), sends it to the TCP socket to be handled (executed) in other language and waits until message with same ``uuid`` is received. After data reception it assigns received ``data`` back to the transaction, so other language can interact with transactions same way like :ref:`native Node.js hooks <hooks-nodejs>`.
 
-Language agnostic test suite
-----------------------------
+What is a hooks handler?
+------------------------
 
-Dredd hooks language abstraction bridge comes with `the language agnostic test suite <https://github.com/apiaryio/dredd-hooks-template>`__. It’s written in Gherkin - language for writing `Cucumber <https://github.com/cucumber/cucumber/wiki/A-Table-Of-Content>`__ scenarios and `Aruba CLI testing framework <https://github.com/cucumber/aruba>`__ and it tests your new language handler integration with CLI Dredd and expected behavior from user’s perspective.
+Hooks handler is a process running separately from Dredd, usually started as Dredd's child process using the :option:`--language` option. When Dredd performs testing, it communicates with the hooks handler over TCP socket. The hooks handler runs hooks for each HTTP transaction and lets Dredd know whether something got modified.
 
-What to implement
------------------
 
-If you want to write a hook handler for your language you will have to implement:
+Hooks handler life cycle
+------------------------
 
--  CLI Command runnning TCP socket server
+#. Dredd starts the command given in the :option:`--language` option as its child process (subprocess). Paths to files with hooks given in :option:`--hookfiles` are resolved to absolute paths and given to the child process as arguments.
+#. The hooks handler reads paths to hooks from arguments and loads the hooks code.
+#. The hooks handler opens a TCP socket on port 61321.
+#. Dredd waits for a moment and then tries to connect to ``http://127.0.0.1:61321``. In case the connection is not successful, it exits with status code 3.
+#. For each :ref:`type of hooks <types-of-hooks>` Dredd creates a message and sends it to the socket. The message contains UUID and serialized :ref:`transaction object <transaction>` (or an array of them, in case of `beforeAll`, `afterAll`). Individual messages are sent as UTF-8 JSON documents separated by a newline.
```

Comment:
To add to this, perhaps I'd specify clearer on "newline". I'm assuming ASCII line feed (hex: `0A`) as the seperator. You mentioned JSON encoding specifically, but didn't mention the encoding of the separator which can be left for different interpretation.

Reply:
Should we clarify the endianness too?

Reply:
Is there ever a need to send an actual newline inside the JSON payload? is there a way to escape newlines?

Reply:
Okay, I know JSON is by default UTF-8, but didn't know it can be other UTFs for some corner cases. My intention behind explicitly mentioning UTF-8 was to address https://github.com/apiaryio/dredd/issues/325. I wonder if the issue can be just closed then, because saying it is JSON should be enough... and if anyone isn't sure, they can always Google out the closed issue for a reference 😄

Reply:
I think properly formed JSON should escape newline characters, so having a newline character as a separator shouldn't be an issue. At least it never was, except for cases when someone feeded the TCP socket with some crazy binary data, but that's not a supported scenario 🙂

Reply:
It's JSON and JSON defines all the encoding it needs so I would not call it out either. Most people will just use whatever JSON package comes with the language and is most likely correct. Given we deal with JSON there is no need for endianess either as everything is UTF-8 string and thus no multibyte stuff where order matters. Correct me if I miss something as I am not a specialist in JSON encoding stuff.